### DISCUSSION TITLE
x-model doesn't work in the CSP build

### DISCUSSION THREAD

**Discussion by akhan619** [alpinejs/alpine#1849] [7. Bugs (archived)]
### Discussed in https://github.com/alpinejs/alpine/discussions/1847

<div type='discussions-op-text'>

<sup>Originally posted by **akhan619** July 25, 2021</sup>
# #Summary
x-model is not reactive when used in the CSP build.

## Test
I cloned the repo and did both ```npm install && npm run build```. No issues there. From there included the ```module.esm.js``` from the csp folder in my app.js and set it up as follows
```
import Alpine from './alpinejs/csp/module.esm'

window.Alpine = Alpine

document.addEventListener('alpine:init', () => {
    Alpine.data('data', () => ({
        message: ''
    }))
})

window.Alpine.start()
```

Tested it with the example in the documentation
```
    <div x-data="data">
        <input type="text" x-model="message">
        <span x-text="message"></span>
    </div>
```

This is not working. As a test used the normal build i.e. non-csp and everything works as expected. Side-note: Any initial value set for ```message``` does reflect in both the input and span. However, they aren't reactive.

## Questions
- Is the CSP build complete? 
- Am I making a mistake in assuming that the we can bind to the Alpine.data() object as I did?

Any feedback is appreciated.</div>

**Comment by domtra** [comment]
@akhan619 I ran into the same issue. Do the changes from [this pr](https://github.com/alpinejs/alpine/pull/2165) fix the issues for you?

**Comment by atevington-centralreach** [comment]
still an issue in current `main` branch

**Comment by maersu** [comment]
I was also not able to make the CSP build working with `x-model` (@alpinejs/csp@3.14.0)

With 

```
<div x-data="data">
	<input x-model="message" type="text"/>
	<span x-text="message"></span>
</div>
```

I get

```
Alpine Error: Alpine is unable to interpret the following expression using the CSP-friendly build:
"message = __placeholder"
Read more about the Alpine's CSP-friendly build restrictions here: https://alpinejs.dev/advanced/csp
 <input x-model=​"message" type=​"text">​
```

A workaround is to do it without `x-model` and do more boilerplate stuff using references (via `x-ref`)

**Reply by ekwoka** [reply]
Shouldn't need refs.

Just a `:value` and `@input`. that's all x-model does.

**Comment by johot** [comment]
Yes I just got the same. Like @ekwoka pointed out I fixed it by doing:

```html
<input x-bind:value="username" x-on:input="setUsername">
```
